common/appscale/common/service_stats/stats_manager.py

```diff
@@ -0,0 +1,492 @@
+    self._history_size = history_size
+    self._force_clean_after = force_clean_after
+    self._last_autoclean_time = now
+
+    # Configure cumulative counters
+    self._start_time = now
+    self._cumulative_counters_config = cumulative_counters
+    self._cumulative_counters = {}
+    _fill_zero_counters_dict(cumulative_counters, self._cumulative_counters)
+
+    # Configure metrics for recent requests
+    self._metrics_for_recent_config = default_metrics_for_recent
+
+  @property
+  def service_name(self):
+    """ Name of service """
+    return self._service_name
+
+  @property
+  def current_requests(self):
+    """ Number of currently running requests """
+    return len(self._current_requests)
+
+  @staticmethod
+  def generate_request_model(request_fields):
+    """ Defines class describing request information.
+
+    Args:
+      request_fields: a list of string names of fields.
+    Returns:
+      a python class having request fields as instance attributes.
+    """
+    for request_field in ServiceStats.RESERVED_REQUEST_FIELDS:
+      if request_field in request_fields:
+        raise ReservedRequestField(
+          "custom_request_fields or request_fields list contains "
+          "reserved name '{}'", format(request_field))
+
+    class RequestInfo(object):
+      __slots__ = set(list(request_fields) +
+                      list(ServiceStats.RESERVED_REQUEST_FIELDS))
+
+      def __init__(self, **fields_dict):
+        # Make sure that new object has all request fields
+        for field in self.__slots__:
+          setattr(self, field, None)
+        self.update(fields_dict)
+
+      def update(self, fields_dict):
+        for field, value in iteritems(fields_dict):
+          try:
+            setattr(self, field, value)
+          except AttributeError as e:
+            raise UnknownRequestField(str(e))
+
+      def __getattr__(self, field):
+        try:
+          object.__getattribute__(self, field)
+        except AttributeError as e:
+          raise UnknownRequestField(str(e))
+
+    return RequestInfo
+
+  def start_request(self, **request_info_dict):
+    """ Adds request to a collection of currently running requests.
+
+    Args:
+      request_info_dict: a dictionary containing initial request info.
+    Returns:
+      an internal request ID which should be used for finishing request.
+    """
+    now = _now()
+    # Instantiate a request_info object and fill its start_time
+    new_request = self._request_info_class(**request_info_dict)
+    new_request.start_time = now
+    # Add currently running request
+    self._last_request_no += 1
+    self._current_requests[self._last_request_no] = new_request
+
+    if now - self._last_autoclean_time > self.AUTOCLEAN_INTERVAL:
+      # Avoid memory leak even if client sometimes don't finish requests
+      self._clean_outdated()
+
+    return self._last_request_no
+
+  def finalize_request(self, request_no, **new_info_dict):
```

Comment:
In my opinion, it would be useful if this were a bit more restricted. If, for example, the necessary details were passed to `start_request`, this class could be turned into a context manager that would be easier to use.

I can see that response code won't be known until after request finishes, but something like a tornado request object could be passed to the init, and the finalize could take care of reading the status. Are there other fields that would complicate this?